src/crimson/os/seastore/seastore_types.cc

```diff
@@ -130,6 +130,212 @@ std::ostream &operator<<(std::ostream &out, const laddr_offset_t &laddr_offset)
 	     << "+" << std::hex << laddr_offset.get_offset() << std::dec;
 }
 
+std::ostream &operator<<(std::ostream &out, const laddr_printer_t &p) {
+  out << p.addr;
+  if (p.addr.is_object_address()) {
+    out << std::hex << "(shard=" << static_cast<int>(p.addr.get_shard())
+	<< ", pool=" << p.addr.get_pool()
+	<< ", reversed_hash=" << p.addr.get_reversed_hash()
+	<< ", local_object_id=" << p.addr.get_local_object_id()
+	<< ", local_clone_id=" << p.addr.get_local_clone_id()
+	<< ", is_metadata=" << p.addr.is_metadata()
+	<< ", block_offset=" << p.addr.get_block_offset()
+	<< ")" << std::dec;
+  }
+
+  return out;
+}
+
+std::ostream &operator<<(std::ostream &out, const laddr_hint_t &hint) {
+  out << "laddr_hint_t(hint=" << laddr_printer_t{hint.addr} << ", conflict_level=";
+  switch(hint.conflict_level) {
+  case laddr_conflict_level_t::object_content:
+    out << "object_content";
+    break;
+  case laddr_conflict_level_t::local_object_id:
+    out << "local_object_id";
+    break;
+  case laddr_conflict_level_t::local_clone_id:
+    out << "local_clone_id";
+    break;
+  case laddr_conflict_level_t::block_offset:
+    out << "block_offset";
+    break;
+  case laddr_conflict_level_t::never:
+    out << "never";
+    break;
+  default:
+    __builtin_unreachable();
+  }
+  out << ", conflict_policy=";
+  switch(hint.conflict_policy) {
+  case laddr_conflict_policy_t::gen_random:
+    out << "gen_random";
+    break;
+  case laddr_conflict_policy_t::linear_search:
+    out << "linear_search";
+    break;
+  default:
+    __builtin_unreachable();
+  }
+  return out << ")";
+}
+
+namespace {
+struct random_generator_t {
+  static random_generator_t &get() {
+    static thread_local random_generator_t r{};
+    return r;
+  }
+  random_generator_t()
+      : eng(std::rand()),
+        global(0, std::numeric_limits<uint64_t>::max()) {}
+  std::default_random_engine eng;
+  std::uniform_int_distribution<uint64_t> global;
+  uint64_t operator()() { return global(eng); }
+};
+uint64_t rand_field() { return random_generator_t::get()(); }
+} // namespace
+
+laddr_hint_t gen_global_random_hint() {
+  return {
+    L_ADDR_MIN.with_object_content(rand_field()),
+    laddr_conflict_level_t::object_content,
+    laddr_conflict_policy_t::linear_search
+  };
+}
+
+laddr_hint_t gen_global_onode_hint(
+  shard_t shard,
+  pool_t pool,
+  crush_hash_t crush) {
+  uint64_t value =
+      ((uint64_t)(shard & 0xFF) << 44) |
+      ((uint64_t)(pool & 0xFFF) << 32) |
+      ((uint64_t)crush);
```

Comment:
This hint design follows the original mechanism of generating hint for onode extents. Compared to random allocation, this approach could take advantages of the locality of onodes in the onode tree, e.g. sibling split, the onodes in a same extent are likely from the same pool/PG. I'm not sure whether it could help reduce the probability of conflicts.

Reply:
From an offline discussion, the hint of onode extent might be appropriate to share the same object prefix with the other object extents, like before.